### PR TITLE
twister: rework board enumeration

### LINKED ISSUE #74142: twister does not understand HWMv2 board qualifiers for platform arguments
**Describe the bug**
The `--platform` argument to twister does not understand HWMv2 board qualifiers.

**To Reproduce**
1. `west build -b lpcxpresso55s36/lpc55s36 samples/hello_world/` (this works as expected)
2. `west twister -p lpcxpresso55s36/lpc55s36 -T samples/hello_world/` (this does not work)
3. See error

**Expected behavior**
Twister is able to understand board qualifiers.

**Impact**
Confusion, especially since our shell completions scripts completes the board qualifiers for `west twister -p ...` as well.

**Logs and console output**
```
$ west twister -p lpcxpresso55s36/lpc55s36 -T samples/hello_world/
INFO    - Using Ninja..
INFO    - Zephyr version: v3.6.0-5871-g97c3f87c82b4
INFO    - Using 'zephyr' toolchain.
ERROR   - platform_filter - unrecognized platform - lpcxpresso55s36/lpc55s36
```

**Environment (please complete the following information):**
 - OS: Linux
 - Toolchain: Zephyr SDK
 - Commit SHA: 2871c87aaec004cf4fe8e61a8d15c35b6d8ab677

### PR DESCRIPTION
**Pull request by nashif**
We now use hwmv2 to list boards instead of relying on twister specific
config files.

One yaml file (twister.yaml for now) will have all the data needed for
all possible targets and variations of a board reusing most of the data
where possible and variations can override the top level data.

Twister keeps track of 'aliases' of boards and identifies that for
example native_sim is the same as native_sim/native, so either names
will be possible in both test yaml files or on the command line,
however, the reporting will always use the full name, so no there is no
confusion about what is being tested/built.

- **twister: rework board handling**
- **boards: other: stm32_min_dev: use one twister.yaml**
- **boards: intel_adsp: use one twister.yaml**

Fixes #74142

### CODE REVIEW

**Review by henrikbrixandersen — COMMENTED** [review]
In general, I like where this is going with one `twister.yaml` file per board directory, allowing to special-case the various variants of a board. This should also clear up some of the confusion as to what the board YAML (now: `twister.yaml`) file is used for.

File: boards/arm/fvp_base_revc_2xaemv8a/fvp_base_revc_2xaemv8a_fvp_base_revc_2xaemv8a_smp_ns.yaml
Comment (ithinuel) [inline boards/arm/fvp_base_revc_2xaemv8a/fvp_base_revc_2xaemv8a_fvp_base_revc_2xaemv8a_smp_ns.yaml:4]: Why is this changed?

As far as I understand omitting the SoC when there's only one is permitted:

from https://docs.zephyrproject.org/latest/hardware/porting/board_porting.html#board-terminology

> If a board contains only one single-core SoC, then the SoC can be omitted from the board target. This implies that if the board does not define any board qualifiers, the board name can be used as a board target. Conversely, if board qualifiers are part of the board definition, then the SoC can be omitted by leaving it out but including the corresponding forward-slashes: //.

**Comment by nashif** [comment]
adapted all tests and fixed many issues discovered by the twister testsuite...

**Comment by nashif** [comment]
@gchwier @golowanow @PerMac  @LukaszMrugala  @hakehuang  please review. it is still in early draft with many cleanups to be done, but the basic functionality is working.

**Comment by hakehuang** [comment]
can we have a yaml processing script that can auto merge yaml files that are impacted to twister.yaml?

**Comment by nashif** [comment]
> can we have a yaml processing script that can auto merge yaml files that are impacted to twister.yaml?

we can, for majority of boards this will not be needed as it means a simple rename only though.

**Comment by nashif** [comment]
added possibility to use the qualifier as a platform key now.

**Comment by rettichschnidi** [comment]
I'd love to see a filename other than `twister.yaml` to be used.

Reason: I still hope that, one day, we can use `twister.yaml` in place of `sample.yaml` and `testcase.yaml`. See also https://github.com/zephyrproject-rtos/zephyr/issues/70223#issuecomment-1997481268.

**Comment by nashif** [comment]
> I'd like to see a filename other than `twister.yaml` to be used.
> 
> Reason, part 1: I still hope that, one day, we can use `twister.yaml` in place of `sample.yaml` and `testcase.yaml`. See also [#70223 (comment)](https://github.com/zephyrproject-rtos/zephyr/issues/70223#issuecomment-1997481268).
> 
> Reason, part 2: I'd love to use Twister to build my (downstream, release) firmwares. Having to use files named `sample.yaml` and `testcase.yaml` feels weird. `twister.yaml` however would be a great name IMHO. But not anymore once the name has already been used for other purposes.

I am planning to replace `sample.yaml` and `testcase.yaml` with one file, maybe `tests.yaml` or something along those lines, not 'twister.yaml'. Main reason here is that those files can also be used by west or other tools, not only twister.

**Comment by butok** [comment]
Hi @nashif
FYI
Some examples use sample.yaml to select a configuration, not for testing. Please. look at this discussion https://github.com/zephyrproject-rtos/zephyr/pull/76222#discussion_r1689251524

File: boards/nxp/mimxrt1060_evk/board.yml
Comment (decsny) [inline boards/nxp/mimxrt1060_evk/board.yml:1]: FYI i'm changing this one in another PR #79517

**Review by hakehuang — DISMISSED** [review]
need fix the CI failure, others looks ok

**Comment by nashif** [comment]
> need fix the CI failure, others looks ok

CI issues are pre-existing

**Review by henrikbrixandersen — DISMISSED** [review]
As mentioned earlier, I really like where this is going. The basic work done here looks great, outstanding work to add `twister.yaml` files to all boards can happen just before merge to avoid conflicts piling up.

**Comment by PerMac** [comment]
I am doing a review and already have some comments that I would like to be addressed but I want to finish reviewing before publishing. Please, don't merge until then and give me one more day or dismiss my "DNM" if I fail to do this in time.

**Comment by nashif** [comment]
> I am doing a review and already have some comments that I would like to be addressed but I want to finish reviewing before publishing. Please, don't merge until then and give me one more day or dismiss my "DNM" if I fail to do this in time.

please do not put DNM on PRs without a good reason., if you have any feedback, use the 'request changes`. This PR has been open for long enough.

**Comment by PerMac** [comment]
I have a draft of some review comments already in github and don;t want to publish in that state, and I believe doing "request change" would publish them as they are now

**Comment by nashif** [comment]
> I have a draft of some review comments already in github and don;t want to publish in that state, and I believe doing "request change" would publish them as they are now

if there is anything that is worth 'request change', post it.  if you put DNM, that is very open-ended and does not tell me anything about why it is DNM. We have a release coming up, and we need to get this resolved.

**Comment by nashif** [comment]
> I have a draft of some review comments already in github and don;t want to publish in that state, and I believe doing "request change" would publish them as they are now

note that I was planning to merge this today, but can wait for one more day, already thrilled about that review that is taking so long to author.

File: scripts/pylib/twister/twisterlib/platform.py
Comment (PerMac) [inline scripts/pylib/twister/twisterlib/platform.py:94]: this is the single place where `board` is used. maybe pass just `board.vendor` in the method definition? It would be less confusing  when one don't have to figure what's the difference between `data` and `board`. Or add `vendor` entry to twister.yaml so everything is in one place?

File: scripts/pylib/twister/twisterlib/twister_main.py
Comment (PerMac) [inline scripts/pylib/twister/twisterlib/twister_main.py:134]: code duplication, the same method already exists in testplan.py

File: scripts/pylib/twister/twisterlib/testplan.py
Comment (PerMac) [inline scripts/pylib/twister/twisterlib/testplan.py:447]: this definition comes in the middle of the `for` loop. please, move it elsewhere

File: scripts/pylib/twister/twisterlib/testplan.py
Comment (PerMac) [inline scripts/pylib/twister/twisterlib/testplan.py:712]: duplicated in twister_main.py

**Review by PerMac — CHANGES_REQUESTED** [review]
Comments marked with
"Can be refactored later on, since it was already like this" can be ignored. I realized after already making some of those that those issues weren't introduced in this PR, hence no need to fix them here. And since I already made those comments I found it worth sharing if you'd be in the mood for also taking care of them now.

**Comment by PerMac** [comment]
> note that I was planning to merge this today, but can wait for one more day, 

Sorry, I know you had it opened for a while. I also understand that this is an important and anticipated improvement. I wasn't able to find time to review it earlier. And since it is important and impactful change I thought you will appreciate extra pair of eyes on it:) 

> already thrilled about that review that is taking so long to author.

The changes are big enough that even github cannot handle showing it fully ;)

File: scripts/pylib/twister/twisterlib/platform.py
Comment (nashif) [reply scripts/pylib/twister/twisterlib/platform.py:59]: board = baord object as parse by list_boards
target = https://docs.zephyrproject.org/latest/glossary.html#term-board-target

I want to have board here, even though I use only one item from that object now, more will follow, for example descriptive name etc as we move more things into board.yml file. 

will add a docstring.

File: scripts/pylib/twister/twisterlib/platform.py
Comment (nashif) [reply scripts/pylib/twister/twisterlib/platform.py:94]: more usage from the board object will follow.

**Comment by nashif** [comment]
> Comments marked with
> "Can be refactored later on, since it was already like this" can be ignored. I realized after already making some of those that those issues weren't introduced in this PR, hence no need to fix them here. And since I already made those comments I found it worth sharing if you'd be in the mood for also taking care of them now.

address many if not all of the comments....

File: scripts/pylib/twister/twisterlib/platform.py
Comment (gchwier) [inline scripts/pylib/twister/twisterlib/platform.py:69]: data is now used before checking if is not None:
```
>>> data = None
>>> data.get("variants", {})
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: 'NoneType' object has no attribute 'get'
```
initialize data before use:
```
    if data is None:
        data = {}
```
or do not use a default value in funcion parameters
(see @PerMac comment https://github.com/zephyrproject-rtos/zephyr/pull/77250#discussion_r1804821953)

Comment (nashif) [reply scripts/pylib/twister/twisterlib/platform.py:69]: addressed

File: scripts/pylib/twister/twisterlib/twister_main.py
Comment (nashif) [reply scripts/pylib/twister/twisterlib/twister_main.py]: addressed

File: scripts/pylib/twister/twisterlib/testplan.py
Comment (nashif) [reply scripts/pylib/twister/twisterlib/testplan.py:714]: addressed

File: scripts/pylib/twister/twisterlib/platform.py
Comment (nashif) [reply scripts/pylib/twister/twisterlib/platform.py]: addressed

File: scripts/pylib/twister/twisterlib/platform.py
Comment (nashif) [reply scripts/pylib/twister/twisterlib/platform.py:67]: addressed

**Review by PerMac — APPROVED** [review]

File: boards/arm/fvp_base_revc_2xaemv8a/fvp_base_revc_2xaemv8a_fvp_base_revc_2xaemv8a_smp_ns.yaml
Comment (golowanow) [inline boards/arm/fvp_base_revc_2xaemv8a/fvp_base_revc_2xaemv8a_fvp_base_revc_2xaemv8a_smp_ns.yaml:4]: looks like this change is included into a wrong commit among others in this PR

File: scripts/pylib/twister/twisterlib/platform.py
Comment (golowanow) [inline scripts/pylib/twister/twisterlib/platform.py:69]: I suppose the related documentation update is planned soon, right ?
https://docs.zephyrproject.org/latest/develop/test/twister.html#board-configuration

File: tests/arch/arm64/arm64_smc_call/testcase.yaml
Comment (golowanow) [inline tests/arch/arm64/arm64_smc_call/testcase.yaml:1]: looks like this change is included into a wrong commit among others in this PR

File: boards/intel/adsp/twister.yaml
Comment (nashif) [reply boards/intel/adsp/twister.yaml:6]: twister.yaml data will apply to all variants and boards unless specified otherwise, no need to add each variant and target if they share the same data.

**Review by henrikbrixandersen — APPROVED** [review]

File: boards/others/stm32_min_dev/board.yml
Comment (nordicjm) [inline boards/others/stm32_min_dev/board.yml:7]: This is wrong, this should never have been made the default, these are completely different boards by completely different vendors that have no relation to one above and there should be no default

Comment (henrikbrixandersen) [reply boards/others/stm32_min_dev/board.yml:7]: Can you submit a follow-up PR showing how you mean it should have been done?

Comment (nordicjm) [reply boards/others/stm32_min_dev/board.yml:7]: It should be reverted, it should just not have a default, you build for stm32_min_dev@blue or stm32_min_dev@black, specifying the revision

Comment (nordicjm) [reply boards/others/stm32_min_dev/board.yml:7]: I see there is a default in the revisions file from the hwmv2 conversion, that should not have been added

**Comment by PerMac** [comment]
FYI with this PR some quarantine file (if used) might have to be updated. Since twister now resolves a platform name to a full one, including revisions (if present), one has to also reflect this in quarantine to get a match. E.g. nrf9160dk/nrf9160 is now resolved to nrf9160dk@0.14.0/nrf9160 and one needs to update their quarantines to this name or use regexed one nrf9160dk.*/nrf9160

File: boards/intel/adsp/twister.yaml
Comment (golowanow) [reply boards/intel/adsp/twister.yaml:6]: #80564

**Pull request merged**
